source/production/F0.Minesweeper.Logic/Minefield.cs

```diff
@@ -63,11 +53,54 @@ public IGameUpdateReport Uncover(Location location)
 				isFirstUncover = false;
 			}
 
-			// TODO: implement real logic
-			return new GameUpdateReport(GameStatus.InProgress,
-				new IUncoveredCell[] {
-					new UncoveredCell(new Location(location.X, location.Y), false, 1)
-				});
+			List<Cell> allUncoveredCells = UncoverCells(location);
+
+			GameStatus gameStatus = GetGameStatus();
+
+			return new GameUpdateReport(gameStatus, allUncoveredCells.ToArray());
+		}
+
+		private GameStatus GetGameStatus()
+		{
+			if (minefield.Any(cell => cell.IsMine && cell.Uncovered))
+			{
+				return GameStatus.IsLost;
+			}
+			if (minefield
+				.Where(cell => !cell.IsMine)
+				.All(nonMineCell => nonMineCell.Uncovered))
+			{
+				return GameStatus.IsWon;
+			}
+			return GameStatus.InProgress;
+		}
+
+		private List<Cell> UncoverCells(Location location)
+		{
+			var returnCellList = new List<Cell>();
+			var uncoveredCell = minefield.Single(cell => cell.Location == location);
+
+			if (uncoveredCell.Uncovered)
+			{
+				return returnCellList;
+			}
+
+			uncoveredCell.Uncovered = true;
+			returnCellList.Add(uncoveredCell);
+
+			if (uncoveredCell.IsMine || uncoveredCell.AdjacentMineCount > 0)
+			{
+				return returnCellList;
+			}
+
+			IEnumerable<Location> locationsAroundCell = Utilities.GetLocationsAreaAroundLocation(AllLocations, location, true);
+
+			foreach (var l in locationsAroundCell)
+			{
+				returnCellList.AddRange(UncoverCells(l));
+			}
+
+			return returnCellList;
```

Comment:
The list that we return here is potentially not distinct in case we have at least two cells beside each other with zero adjacent mines. I don't think that's a good idea. To solve that I'd prefer if we'd use a `HashSet<Cell>` instead of a `List<Cell>`

Reply:
I don't think that we ever return here a list with duplicates.
With two adjacent cells with value 0 the flow would be as follows:
* The first cell is being uncovered and uncovers all neighbor cells
* The second cell is then uncovered and also wants to uncover all its neighbors
* This will call `UncoverCells` on the first cell again but as this cell is already uncovered it will return an empty list in the first `if` of the method.

But I would still like to improve my UncoverCells method and I think that using `HashSet<Cell>` can be valuable 🤔

Reply:
My mind is just too stupid. You're totally right that here no duplicates will occur. I stand corrected :D 
But then I wouldn't use a `HashSet` since this would add an unnecessary complex construct with no purpose behind it.